### PR TITLE
Implement a new resource claim controller pattern

### LINKED ISSUE #119: Update connection secret naming override in databases to pattern in buckets
Migrate databases to override the connection secret in the same way that the buckets do. 

Currently in the databases spec there is a connectionSecretRef that can either act as an override name or is set with a default name after the database has been initialized. The reason for the change is to not have the operator need to conditionally update the Spec after initialization and instead only update the status.

The bucket pattern is to move ConnectionSecretRef into status, and to add a override to Spec like the following:
```go
S3BucketSpec struct {
...
ConnectionSecretNameOverride string `json:"connectionSecretNameOverride,omitempty"`
}

S3BucketStatus struct {
...
ConnectionSecretRef v1.LocalObjectReference `json:"connectionSecretRef,omitempty"`
}
```

### LINKED ISSUE #167: Investigate common resource fields to simplify code
I was also thinking today it could be very nice to add something like the following. We only comment in one place and as we make additions we could make them in a single place, and any new ones would get the right fields. 

```go
type ClassedResource struct {
        // ClaimRef reference to the Generic Class Type
	ClaimRef    *corev1.ObjectReference     `json:"claimRef,omitempty"`
        // ClassRef reference to the ResourceClass for this Resource
	ClassRef    *corev1.ObjectReference     `json:"classRef,omitempty"`
 }
      
type ProviderResource struct {
        // ProviderRef reference to the provider details
	ProviderRef corev1.LocalObjectReference `json:"providerRef"`
        Policy      `json:",inline"`
        // ConnectionSecretNameOverride set this override the generated name of Status.ConnectionSecretRef.Name
	ConnectionSecretNameOverride string `json:"connectionSecretNameOverride,omitempty"`
}
```

_Originally posted by @lukeweber in https://github.com/_render_node/MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDIzNjkwOTIwMQ==/comments/review_comment_

### LINKED ISSUE #168: Standardize connection secrets
Connection Secrets are very similar, but there are small variations across the codebase. Currently, we dynamically set the LocalObjectReference if none was set inside the Spec. 

I would like to standardize on the following:
* Set ConnectionSecretRef in Status, not in Spec
* Add an override field in Spec to make ConnectionSecretName deterministic when needed
* else Default to the name of the object + a random suffix if no name is set. Default should make sure to not overflow max object length and trim the name if necessary.

_Originally posted by @lukeweber in https://github.com/crossplaneio/crossplane/pull/150_

### LINKED ISSUE #440: Remove "selector" from Resource Claim specs
https://github.com/crossplaneio/crossplane/blob/master/pkg/apis/compute/v1alpha1/types.go#L32

I think the selector was added by mistake. Currently, it is not used and I don't see any use cases for this selector at this level.

### LINKED ISSUE #450: Resource claims ignore controller ref when creating connection secrets
**Is this a bug report or feature request?** Bug Report

**Deviation from expected behavior:**
Assume a `PostgresqlInstance` kind resource claim named `cool` currently exists. You then create a `RedisCluster` kind resource claim with the same name in the same namespace. The `RedisCluster`'s connection secret will overwrite the `PostgresqlInstance`'s connection secret.

**Expected behavior:**
The second resource claim should fail with a conditioned status message explaining that a connection secret named `cool` already exists and is owned by another resource claim. We already set a controller reference when creating resource claim connection secrets, but do not check whether we own an existing secret before updating/overwriting it.

### LINKED ISSUE #451: Require a standard set of Resource Claim spec fields
Currently most (all?) resource claim types include a `ClassRef`, `ResourceRef`, and a `Selector` field.

```go
type RedisClusterSpec struct {
	ClassRef    *corev1.ObjectReference `json:"classReference,omitempty"`
	ResourceRef *corev1.ObjectReference `json:"resourceName,omitempty"`
	Selector    metav1.LabelSelector    `json:"selector,omitempty"`
}
```

The latter is seemingly unused per https://github.com/crossplaneio/crossplane/issues/440, while the former two are effectively mandatory as they're required to satisfy the `core.ResourceClaim` interface:
```go
type ResourceClaim interface {
	runtime.Object
	metav1.Object
	ClaimStatus() *ResourceClaimStatus
	OwnerReference() metav1.OwnerReference
	ObjectReference() *corev1.ObjectReference
	ClassRef() *corev1.ObjectReference
	ResourceRef() *corev1.ObjectReference
	SetResourceRef(*corev1.ObjectReference)
}
```

A few resource claims, including `Bucket`, also support the `ConnectionSecretNameOverride` field. This field is particularly useful to allow two resource claims of different kinds to have the same namespace and name without fighting over their connection secret, which by default is named identically to the claim, per https://github.com/crossplaneio/crossplane/issues/450.
```go
	ConnectionSecretNameOverride string `json:"connectionSecretNameOverride,omitempty"`
```

I suggest we encode this standard by creating a new type alongside the existing `corev1alpha1.ResourceClaimStatus`: `corev1alpha1.ResourceClaimSpec`:

```go
type ResourceClaimSpec struct {
	ClassRef    *corev1.ObjectReference `json:"classReference,omitempty"`
	ResourceRef *corev1.ObjectReference `json:"resourceName,omitempty"`

	ConnectionSecretOverride *corev1.LocalObjectReference `json:"connectionSecretOverride,omitempty"`
}
```

While most (all?) resource claims use `ResourceClaimStatus` itself to represent their status, I expect `ResourceClaimSpec` would be embedded in a more targeted spec struct to allow us to support the extra fields a spec requires, for example:

```go
type PostgreSQLInstanceSpec struct {
	corev1alpha1.ResourceClaimSpec `json:",inline"`

	// +kubebuilder:validation:Enum=9.6
	EngineVersion string `json:"engineVersion,omitempty"`
}
```

### LINKED ISSUE #456: Remove workload.compute.crossplane.io?
https://github.com/crossplaneio/crossplane/pull/420 added support for complex workloads in the `workload.crossplane.io` API namespace, but left the original `Workload` resource and controllers under the `compute.crossplane.io` namespace.

We should probably remove the older `Workload` implementation. Much of our documentation and examples target this older implementation. We should update them to target complex workloads before we remove the old `Workload`.

### LINKED ISSUE #477: ResourceController Thrashing on Wait
ResourceController is triggering perpetual reconcile loop:
```
2019-05-19T10:26:45.630Z        DEBUG   crossplane.controller.s3bucket.aws.crossplane.io        reconciling     {"kind": "s3bucket.storage.aws.crossplane.io/v1alpha1", "request": "crossplane-system/bucket-faccada4-7a1f-11e9-bd86-42010a9a01bd"}
2019-05-19T10:26:46.180Z        DEBUG   crossplane.controller.s3bucket.aws.crossplane.io        reconciling     {"kind": "s3bucket.storage.aws.crossplane.io/v1alpha1", "request": "crossplane-system/bucket-faccada4-7a1f-11e9-bd86-42010a9a01bd"}
2019-05-19T10:26:46.739Z        DEBUG   crossplane.controller.s3bucket.aws.crossplane.io        reconciling     {"kind": "s3bucket.storage.aws.crossplane.io/v1alpha1", "request": "crossplane-system/bucket-faccada4-7a1f-11e9-bd86-42010a9a01bd"}
2019-05-19T10:26:47.282Z        DEBUG   crossplane.controller.s3bucket.aws.crossplane.io        reconciling     {"kind": "s3bucket.storage.aws.crossplane.io/v1alpha1", "request": "crossplane-system/bucket-faccada4-7a1f-11e9-bd86-42010a9a01bd"}
```

This appears due to perpetual status update:
```
   - LastTransitionTime: "2019-05-19T10:28:18Z"
      Message: Resource is not in running state
      Reason: Waiting for resource to become available
      Status: "True"
      Type: Pending
```
...
```
    - LastTransitionTime: "2019-05-19T10:28:48Z"
      Message: Resource is not in running state
      Reason: Waiting for resource to become available
      Status: "True"
      Type: Pending
```
Note: `LastTransitionTime` is being updated every single time.

This is due to: https://github.com/crossplaneio/crossplane/blob/master/pkg/controller/core/resource_controller.go#L190

When we `UnsetAllCondition` and then activate a specific condition, we change the activated condition timestamp. Instead, we should check if condition has changed and only then unset all conditions if needed, otherwise we should leave current condition intact.

### LINKED ISSUE #504: Refactor all resources to use the new ConditionedStatus
In https://github.com/crossplaneio/crossplane/issues/495 we arrived at a new design for the `ConditionedStatus` that each Crossplane resource embeds in its `Status` struct. This design was introduced to the codebase in https://github.com/crossplaneio/crossplane/pull/499. In order to break this massive refactoring across multiple PRs we renamed the old `ConditionedStatus` to `DeprecatedConditionedStatus`. This should allow us to port each resource over to the new pattern one over time.

This issue should be considered resolved once `DeprecatedConditionedStatus` has been removed from the codebase.

### LINKED ISSUE #518: Require explicit connection secret references
Currently both Crossplane resources and resource claims create connection secrets - a Kubernetes `Secret` that stores details required to consume a managed resource such as connection URI, username, password, etc.

There's very little consistency in the implementations. Most resources allow the user to specify a secret name in `.spec` but will generate a secret name if said name is omitted. This generated name is typically just the resource's name. Some resources record the generated name by updating `.spec`, while others record the generated name by updating `.status`.

All resource claims write their secret name to `.status.credentialsSecret`. Most claims simply name their secrets after themselves, but some support fields like `.connectionSecretOverride` to specify an explicit name.

I propose we do away with all this inconsistency and complication, and _require_ both resources and claims to include a `corev1.LocalObjectReference` at `.spec.writeConnectionSecret`. This avoids any surprise or ambiguity for the user about where connection secrets will be written, and simplifies the logic of our applications.

### LINKED ISSUE #547: Redesign the core resource claim controller
From @negz description in #543:

The core controller is responsible for satisfying resource claims by provisioning and/or binding concrete resources - e.g. it satisfies a `RedisCluster` claims by creating a `CloudMemorystoreInstance` resource.

The “key features” of this proposed pattern are:

* Improved requeue logic for all resource claims. Concrete resource controllers must queue a new reconcile speculatively in case the resource they manage changes in the cloud, because there’s no way to subscribe to changes. Resource claims on the other hand can watch the managed resource (i.e. Kubernetes API Resource) they bind to. This means we never need to requeue reconciles in the happy path; we’ll be notified if a claim or resource changes. This relates to #241 and #447.

* All claims and resources use the newer conditioned status added in #447 and #514. These statuses are less confusing (a resource can’t be failed and ready at the same time), and discourage making decisions based on the status rather than the real world.

* All claims and resources support a set of standard fields (claim reference, class reference, etc) by embedding a struct that defines said fields (fixes #451). One of those fields is `.spec.writeConnectionSecretTo`, a consistent and unsurprising field that specifies where a claim or resource should write their connection secret. This replaces the inconsistent support for `.spec.connectionSecretNameOverride` and the surprising behaviour of writing a connection secret named for the resource (fixes #518 and #168). The proposed controller also sets respects controller references when managing secrets (fixes #450).

Another notable change is to the API exposed to users of the “core controller”. In the contemporary implementation the core controller is designed to be embedded in a reconciler for a specific kind of claim, passing any resource-specific logic to a set of handlers for a specific kind of resource. 

In this proposed design the core controller itself is the reconciler; it need not be embedded or wrapped in any reconciler logic. There’s also a 1:1 claim:resource reconciler relationship; i.e. a resource claim kind that can be satisfied by three different resource kinds would use three instances of the core controller, rather than one core controller with three handlers. Instantiating the proposed core controller is typically a case of defining a resource-specific `Configurator` function that populates a spec given a class and claim, then writing a small `Add` function to instantiate the core reconciler, setup watches, and add it to the manager. The `NewReconciler` function also supports plugging in different CRUD logic in case a particular claim kind wants to override the defaults. One (arguably) nice property of this pattern is that the controller that binds a resource claim to a concrete resource can live in the same package as the concrete resource’s controller.

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes
The core controller is responsible for satisfying resource claims by provisioning and/or binding concrete resources - e.g. it satisfies a `RedisCluster` claims by creating a `CloudMemorystoreInstance` resource.

The “key features” of this proposed pattern are:

* Improved requeue logic for all resource claims. Concrete resource controllers must queue a new reconcile speculatively in case the resource they manage changes in the cloud, because there’s no way to subscribe to changes. Resource claims on the other hand can watch the managed resource (i.e. Kubernetes API Resource) they bind to. This means we never need to requeue reconciles in the happy path; we’ll be notified if a claim or resource changes. This relates to #241 and #447.

* All claims and resources use the newer conditioned status added in #447 and #514. These statuses are less confusing (a resource can’t be failed and ready at the same time), and discourage making decisions based on the status rather than the real world.

* All claims and resources support a set of standard fields (claim reference, class reference, etc) by embedding a struct that defines said fields. One of those fields is `.spec.writeConnectionSecretToRef`, a consistent and unsurprising field that specifies where a claim or resource should write their connection secret. This replaces the inconsistent support for `.spec.connectionSecretNameOverride` and the surprising behaviour of writing a connection secret named for the resource. The proposed controller also sets and respects controller references when managing secrets.

Another notable change is to the API exposed to users of the “core controller”. In the contemporary implementation the core controller is designed to be embedded in a reconciler for a specific kind of claim, passing any resource-specific logic to a set of handlers for a specific kind of resource. In this proposed design the core controller itself is the reconciler; it need not be embedded or wrapped in any reconciler logic. There’s also a 1:1 claim:resource reconciler relationship; i.e. a resource claim kind that can be satisfied by three different resource kinds would use three instances of the core controller, rather than one core controller with three handlers.

Instantiating the proposed core controller is typically a case of defining a resource-specific `Configurator` function that populates a spec given a class and claim, then writing a boilerplate `AddClaim` function to instantiate the core reconciler, setup watches, and add it to the manager. The `NewClaimReconciler` function supports plugging in different CRUD logic in case a particular kind wants to override the defaults. An example of this is the `ManagedBinder` interface that updates the binding status of managed resources and their claims; `NewClaimReconciler` defaults to an `APIManagedBinder` which calls `client.Update()` on managed resource, but can easily be asked to use a `APIStatusManagedBinder` which instead calls `client.Status().Update()` for managed resources that use the status subresource.

### Design changes
This PR makes a few design changes of note:

* All managed resources communicate their readiness for binding to a claim by changing their binding phase from “unbindable” to “unbound”. “Unbound” managed resources are considered to be unused but ready for binding.
* All `fooReference` fields are renamed to `fooRef` for consistency. We previously used a mix of both.

### Known issues
Removing speculative requeueing from the resource claim controllers uncovered a race when writing and propagating connection secrets. Some managed resource types take a handful of requeues in order to fully populate their connection secrets. This was not an issue when the resource claim controller continually requeued, copying the managed resource connection secret to the claim connection secret each time resulting in eventual consistency.

Now that the resource claim controller watches managed resources instead of speculatively requeueing it will propagate the managed resource’s connection secret when the managed resource becomes ready for binding, but will not make any further attempts to copy the secret until either the managed resource changes or the global watch sync occurs. This can result in partially populated resource claim connection secrets existing for some time. The Azure SQL controllers and all Redis controllers are affected by this issue. I believe this is a bug in the managed resource controllers, and thus should be fixed in a separate PR.

### Addressed issues
Fixes #547 
Fixes #451 
Fixes #518 
Fixes #168
Fixes #119 
Fixes #450 
Fixes #504
Fixes #477
Fixes #456
Fixes #440
Fixes #167

Partial fix for #298. All resource claim types now use the status subresource.
Partial fix for #241. Resource claims watch their resources instead of requeueing.

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [x] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplaneio/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

**Comment by negz** [comment]
This PR takes us down from ~66% to ~64% code coverage according to Sonar. As best I can tell this is mostly for two reasons:

* We now have more `Add(mgr manager.Manager)` type boilerplate controller registration functions. We historically don't test these functions, and they're still not personally high up my list of logic that I'm concerned about testing.
* I added various getter and setter methods to our API types. I don't think these methods warrant testing.

File: cluster/charts/crossplane/crds/core/v1alpha1/resourceclass.yaml [hunk 1]
```diff
@@ -61,7 +61,7 @@ spec:
           type: string
       required:
       - provisioner
-      - providerRef
+      - providerReference
```
Comment (displague) [inline cluster/charts/crossplane/crds/core/v1alpha1/resourceclass.yaml]: Why was `providerRef` renamed to `providerReference`?

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#object-references

Comment (displague) [reply cluster/charts/crossplane/crds/core/v1alpha1/resourceclass.yaml]: If this was intended, it looks like providerRef -> providerReference changes may have been missed in types.go comments
```console
$ kubectl -n crossplane-system get resourceclass -A
NAMESPACE           NAME               PROVISIONER                                            PROVIDER-REF   RECLAIM-POLICY   AGE
crossplane-system   standard-cluster   gkecluster.compute.gcp.crossplane.io/v1alpha1                         Delete           51m
crossplane-system   standard-mysql     cloudsqlinstance.database.gcp.crossplane.io/v1alpha1                  Delete           51m
```

Comment (negz) [reply cluster/charts/crossplane/crds/core/v1alpha1/resourceclass.yaml]: Good catch on the API conventions. We previously had a mix of `fooRef` and `fooReference`, so I renamed these to standardise. I was unaware the API conventions had a position on this and picked the wrong one.  I've just pushed a commit that renames all the `fooReference` fields to `fooRef`, which should also fix the comments.

**Comment by negz** [comment]
@displague I'll address these inline.

> I ran into this while helm-upgrading my way into the `resourceful` branch. I forced past it, but this might cause others upgrade problems.

I'd rather not fix this; I think this change is large enough and we're alpha enough that it's not unreasonable for the upgrade path to be "completely remove and reinstall it".

> I tried running the AWS Kubernetes Wordpress demo from this branch, only changing `providerRef` to `providerReference` and it appears that the request for the `KubernetesCluster` is not being picked up by anything (it has no `status` and the `crossplane` deployment logs only reflect that it started `eventsource`, `controllers`, and `workers`).
> 
> The GCP example got further, but failed to connect the cluster secret to the `kubernetesapplication`:

I believe both of these were due to my having not updated these examples with `.spec.writeConnectionSecretToRef`, which I now have.

> It looks like the print field `CLUSTER-REF` should use `.spec.resourceReference.name`.

I believe this should be fixed by my having renamed the field back to `.spec.resourceRef`.

**Comment by negz** [comment]
@jbw976 I believe this is now ready for review.

Per discussion with yourself and based on @bassam's input, I've pushed a new commit that reverts to Crossplane's existing relationship between resources and their claims. That is to say, if this PR is merged:

* We'll keep creating resources in the namespace of the resource class (not claim) as we do today.
* We'll continue to set the claim as a (potentially invalid cross-namespace) owner reference of the resource as we do today.
* We'll maintain the same `ReclaimPolicy` semantics as we do today, as detailed in my spirit quest in #550.

Note that we still enqueue a reconcile request for a resource's claim when the resource changes. We do this using the new `*EnqueueRequestForClaim` `handler.EventHandler`, rather than the `*EnqueueRequestForOwner` `handler.EventHandler`.

**Comment by jbw976** [comment]
And just wanted to also add an agreement that Upgrade scenarios are not currently supported while we are in this early alpha phase.  We will eventually need to make changes in a way that either maintains backwards compatibility or performs a migration, but for right now we have the freedom to make breaking API changes.

**Pull request merged**